### ISSUE TITLE
Wildlife guides - Image display

### ISSUE THREAD

**Issue by davidnicholls125** [johnvanbreda/naturespot-d8#184]
Sorry if you've already told me but how do I get images to display in the wildlife guides with the same height? I've copied the code you used in the Meadow Cranesbill example and replaced the image source on the guide entry below:
https://www.naturespot.org.uk/node/251042/edit?destination=/node/251041
Although I resized both images before inserting them to be 150px in height, they are displaying at different heights. I've removed the height and width tags but this hasn't made any difference. 
On a related point, where can I view the CSS set up? I've looked in Bootstrap but can't see it. Thanks.

**Comment by johnvanbreda** [comment]
To check the CSS working for these images:
1. First, ensure the browser developer tools are enabled (see https://balsamiq.com/support/faqs/browserconsole/).
2. Depending on the browser, you should be able to view the page (https://www.naturespot.org.uk/node/251041) then right click on an offending image and select "Inspect". This will open the developer tools on the correct tab (called Elements on Edge - may differ on other browsers but it will do roughly the same thing). 
3. You should be able to see the HTML around the selected image in one pane on the tab:
![image](https://github.com/johnvanbreda/naturespot-d8/assets/2392594/f78f49f5-0f10-49ff-b7ac-708919245a04)
4. You can also see the applied CSS styles in another pane (this is just part of the styling - there is a scrollbar to access more):
![image](https://github.com/johnvanbreda/naturespot-d8/assets/2392594/5c3be0fb-0727-4460-a6c9-bdc9203fd5b6)
5. From this, you can see that the `img` element in the HTML does not specify a width or height. This means it will take it's width or height from the CSS and will revert to the original image width or height only if not specified in the CSS. But the CSS sets the width to 100% and the height to auto, which is a common approach for making an image responsive. This means it will resize itself to fill the width of the containing box. 
6. The containing box in the HTML is defined as `<div class="col-md-3">`. A div is a generic HTML container and giving it a class of the format `col-*-*` means it is a bootstrap column, in this case a 3 column width (out of a grid of 12) column, that switches to stacked mode when below a medium size screen width breakpoint. So, your image will always attempt to fill a 25% width column and the height will resize accordingly.

I vaguely recall a discussion about whether to use media library vs the other image insert tools but can't find it in my emails - do you remember why we said not to use the media library?

**Comment by davidnicholls125** [comment]
Thanks, however specifying a height doesn't seem to have any effect. I've added in the 150px height on the entry below but the display on the guide ignores this. So how do we specify the height? The images I added have been resized to 150 in height but of course they then have variable widths.
https://www.naturespot.org.uk/node/251042/edit?destination=/node/251041
I know you advised using the media library and I have tried it a few times but found it hard work. Is it possible to edit/crop images in the media library?

**Comment by johnvanbreda** [comment]
The height and width setting in the CSS is taking precedence over the height and width attributes you've added to the HTML so I'd have to remove them from the stylesheet. Although I can do that, this means your images are no longer responsive.

If you want the images to be responsive (adapting to different device widths) whilst also being the same height, you will have to ensure your images are the same aspect ratio. At the moment you have a 160 x 150px image and a 200 x 150px image.

**Comment by davidnicholls125** [comment]
Thought this was the case... Is there a way to edit the images in the media library or will this have to be done in another package? Imce allows images to be resized but only in one dimension if you want to keep the aspect ratio.

**Comment by johnvanbreda** [comment]
I don't think having a built in image editor tool for individual images is actually very well supported in Drupal, but there are good reasons for this as the image styles system is able to handle this requirement in a much more consistent and reliable way.

We can create a media style that is able to "scale and crop" - this means it will scale so that the image fits within a defined box size (width and height), then crop off any pixels that are outside the defined size in one direction. The way to do this here would be to create an image style that is either square, or something like a 4:3 aspect ratio, but larger than the actual number of pixels you are expecting to output at. It would actually have to be something like a 700px square or larger, because if you resize your browser and reduce the width using the Groundsel example you will see that the small images you are uploading are actually very pixelated when the 2 image columns stack on top of each other on a mobile device (the third image is added using the new technique so is much better)  So we need the actual image to appear at a reasonably high quality on all devices, which means the 160px x 150px image is much too small.  

I've added 2 image styles, Guide content square and Guide content 4:3. The 3rd Groundsel image is added from the media library using this technique.

Another benefit of this technique is that we can update all images across the site with 1 configuration change, for example you could change all guide images from 4:3 to 16:9 in one go, or if we realise 700px is not big enough for some layouts we can fix it. If you crop the original images using the old technique, then you have to re-do every image. Note though that I do not advise adding images direct from a modern camera into the library as disk space will quickly fill, even though the image style system means that images will be optimised automatically before appearing on pages. I would advise scaling images to the maximum size they are likely to appear on the site at, which I think is 1024px wide, before upload. There are various tools able to batch scale images before you upload them, e.g. Gimp (https://www.thewindowsclub.com/how-to-batch-resize-images-with-gimp-in-windows-10). 

So, in summary:
1) Scale your images to 1024px wide before upload.
2) Add them using the media library button.
3) Use the edit media button to set the image style to Guide content square or Guide content 4:3, using the same format for all images in each guide entry.
4) We can then consider if any other guide content image styles are useful, e.g. a 3:4 size might be good for upright plants.

**Comment by davidnicholls125** [comment]
Thanks John, that all works well. I'm mainly using images from the site so they are already resized to 1024 but will do this myself if uploading others.
